### PR TITLE
feat(pageserver): support schedule gc-compaction

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

part of https://github.com/neondatabase/neon/issues/9114

gc-compaction can take a long time. This patch adds support for scheduling a gc-compaction job. The compaction loop will first handle L0->L1 compaction, and then gc compaction. The scheduled jobs are stored in a non-persistent queue within the tenant structure.

This will be the building block for the partial compaction trigger -- if the system determines that we need to do a gc compaction, it will partition the keyspace and schedule several jobs. Each of these jobs will run for a short amount of time (i.e, 1 min). L0 compaction will be prioritized over gc compaction.

## Summary of changes
 
* Add compaction scheduler in tenant.
* Run scheduled compaction in integration tests.
* Change the manual compaction API to allow schedule a compaction instead of immediately doing it.
* Add LSN upper bound as gc-compaction parameter. If we schedule partial compactions, gc_cutoff might move across different runs. Therefore, we need to pass a pre-determined gc_cutoff beforehand. (TODO: support LSN lower bound so that we can compact arbitrary "rectangle" in the layer map)
* Refactor the gc_compaction internal interface.

### CODE REVIEW

File: pageserver/src/tenant/timeline/compaction.rs [hunk 1]
```diff
@@ -1826,7 +1821,15 @@ impl Timeline {
...
-            let gc_cutoff = gc_info.cutoffs.select_min();
+            let real_gc_cutoff = gc_info.cutoffs.select_min();
+            // The compaction algorithm will keep all keys above the gc_cutoff while keeping only necessary keys below the gc_cutoff for
+            // each of the retain_lsn. Therefore, if the user-provided `compact_below_lsn` is larger than the real gc cutoff, we will use
+            // the real cutoff.
+            let mut gc_cutoff = options.compact_below_lsn.unwrap_or(real_gc_cutoff);
+            if gc_cutoff > real_gc_cutoff {
+                warn!("provided compact_below_lsn={} is larger than the real_gc_cutoff={}, using the real gc cutoff", gc_cutoff, real_gc_cutoff);
+                gc_cutoff = real_gc_cutoff;
+            }
```
Comment (problame) [inline pageserver/src/tenant/timeline/compaction.rs:1832]:
```suggestion
            let gc_cutoff = {
                let real_gc_cutoff = gc_info.cutoffs.select_min();
                // The compaction algorithm will keep all keys above the gc_cutoff while keeping only necessary keys below the gc_cutoff for
                // each of the retain_lsn. Therefore, if the user-provided `compact_below_lsn` is larger than the real gc cutoff, we will use
                // the real cutoff.
                let mut gc_cutoff = options.compact_below_lsn.unwrap_or(real_gc_cutoff);
                if gc_cutoff > real_gc_cutoff {
                    warn!("provided compact_below_lsn={} is larger than the real_gc_cutoff={}, using the real gc cutoff", gc_cutoff, real_gc_cutoff);
                    gc_cutoff = real_gc_cutoff;
                }
```

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -776,6 +776,20 @@ pub(crate) enum CompactFlags {
...
+    pub compact_below_lsn: Option<LsnDisplay>,
+    /// Whether the compaction job should be scheduled.
+    #[serde(default)]
+    pub scheduled: bool,
+}
+
+#[serde_with::serde_as]
+#[derive(Debug, Clone, serde::Deserialize)]
+pub(crate) struct LsnDisplay(#[serde_as(as = "serde_with::DisplayFromStr")] pub Lsn);
+
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:792]: I think this is unnecessary. The `Serialize` impl of `Lsn` already uses the hex repr for human readable formats.

File: test_runner/regress/test_compaction.py [hunk 3]
```diff
@@ -133,20 +145,27 @@ def test_pageserver_gc_compaction_smoke(neon_env_builder: NeonEnvBuilder):
...
             tenant_id,
             timeline_id,
             enhanced_gc_bottom_most_compaction=True,
             body={
-                "start": "000000000000000000000000000000000000",
-                "end": "030000000000000000000000000000000000",
+                "scheduled": True,
+                "compact_range": {
+                    "start": "000000000000000000000000000000000000",
+                    # skip the SLRU range for now -- it races with get-lsn-by-timestamp, TODO: fix this
```
Comment (problame) [inline test_runner/regress/test_compaction.py:156]: What's the GH issue number for this?

**Review by problame — APPROVED** [review]
It would be nice to add a management API for clearing the scheduled compactions list.
I guess it can be achieved through tenant reset API, though.

---

The code in `fn compaction_iteration` is quite messy. I think an abstraction to separate policy (what needs doing next) from the machinery to execute it (calling `compact_with_options`) would be appropriate.

For example:
- comapction_iteration calls that abstraction
- the abstraction yields a`ScheduledCompactionTask` 
- the compaction_iteration invokes `compact_with_options` to execute

WDYT? Doesn't have to be done in tis PR but I'm unwilling to tolerate the messiness in the longer term.
Feel free to create a new issue that references this proposal.

**Comment by skyzh** [comment]
> the abstraction yields aScheduledCompactionTask

Yes I think we need a separation from compact options from compact task. We already have GcCompactionTask in compaction.rs. I'll do a refactor to separate them out after the patches go through.

File: test_runner/regress/test_compaction.py (hunk 3 above)
Comment (skyzh) [reply test_runner/regress/test_compaction.py:156]: https://github.com/neondatabase/neon/issues/10030

**Comment by skyzh** [comment]
I've created an issue to track that and I'll get it done by the next Wednesday to avoid having tech debt in the code

**Pull request merged**